### PR TITLE
Using prior on all camera poses.

### PR DESCRIPTION
**Pull request by ayushbaid**
As we use rotation and translation modules in our pipelines, we have a pretty strong prior on the camera parameters. I think this was suggested by @dellaert quite a while back too.

The metrics do not have any significant difference.

Old:
```
[2021-05-25 14:46:02,542 DEBUG data_assoc.py line 133 198494] [Data association] output number of tracks: 8965
[2021-05-25 14:46:02,542 DEBUG data_assoc.py line 134 198494] [Data association] output avg. track length: 5.23
[2021-05-25 14:46:03,204 INFO bundle_adjustment.py line 48 198495] Input: 8965 tracks on 12 cameras

[2021-05-25 14:46:09,476 INFO bundle_adjustment.py line 113 198495] initial error: 29036343.05
[2021-05-25 14:46:09,477 INFO bundle_adjustment.py line 114 198495] final error: 24937.02
[2021-05-25 14:46:15,056 INFO bundle_adjustment.py line 121 198495] [Result] Number of tracks before filtering: 8965
[2021-05-25 14:46:16,466 INFO bundle_adjustment.py line 129 198495] [Result] Number of tracks after filtering: 8926
[2021-05-25 14:46:16,466 INFO bundle_adjustment.py line 130 198495] [Result] Mean track length 5.230
[2021-05-25 14:46:16,466 INFO bundle_adjustment.py line 131 198495] [Result] Median track length 4.000
[2021-05-25 14:46:16,929 INFO gtsfm_data.py line 308 198495] Min scene reproj error: 0.000
[2021-05-25 14:46:16,929 INFO gtsfm_data.py line 309 198495] Avg scene reproj error: 0.171
[2021-05-25 14:46:16,930 INFO gtsfm_data.py line 310 198495] Median scene reproj error: 0.136
[2021-05-25 14:46:16,930 INFO gtsfm_data.py line 311 198495] Max scene reproj error: 2.949
```

New:
```
[2021-05-25 15:18:25,387 DEBUG data_assoc.py line 133 202930] [Data association] output number of tracks: 8965
[2021-05-25 15:18:25,388 DEBUG data_assoc.py line 134 202930] [Data association] output avg. track length: 5.23
[2021-05-25 15:18:25,852 INFO bundle_adjustment.py line 48 202930] Input: 8965 tracks on 12 cameras

[2021-05-25 15:18:31,823 INFO bundle_adjustment.py line 108 202930] initial error: 29036343.05
[2021-05-25 15:18:31,834 INFO bundle_adjustment.py line 109 202930] final error: 25111.50
[2021-05-25 15:18:36,428 INFO bundle_adjustment.py line 116 202930] [Result] Number of tracks before filtering: 8965
[2021-05-25 15:18:40,280 INFO bundle_adjustment.py line 124 202930] [Result] Number of tracks after filtering: 8926
[2021-05-25 15:18:40,280 INFO bundle_adjustment.py line 125 202930] [Result] Mean track length 5.230
[2021-05-25 15:18:40,280 INFO bundle_adjustment.py line 126 202930] [Result] Median track length 4.000
[2021-05-25 15:18:40,741 INFO gtsfm_data.py line 308 202930] Min scene reproj error: 0.000
[2021-05-25 15:18:40,742 INFO gtsfm_data.py line 309 202930] Avg scene reproj error: 0.181
[2021-05-25 15:18:40,742 INFO gtsfm_data.py line 310 202930] Median scene reproj error: 0.145
[2021-05-25 15:18:40,742 INFO gtsfm_data.py line 311 202930] Max scene reproj error: 2.985
```

### CODE REVIEW

File: gtsfm/bundle/bundle_adjustment.py [hunk 1]
```diff
@@ -68,20 +68,15 @@ def run(self, initial_data: GtsfmData) -> GtsfmData:
...
-        # Also add a prior on the position of the first landmark to fix the scale
-        graph.push_back(
-            gtsam.PriorFactorPoint3(
-                P(0), initial_data.get_track(0).point3(), gtsam.noiseModel.Isotropic.Sigma(POINT3_DOF, 0.1)
+        # Add a prior on all the poses.
+        for i in valid_camera_indices:
+            graph.push_back(
+                gtsam.PriorFactorPinholeCameraCal3Bundler(
+                    C(valid_camera_indices[0]),
+                    initial_data.get_camera(valid_camera_indices[0]),
```
Comment (akshay-krishnan) [inline gtsfm/bundle/bundle_adjustment.py:76]: Please correct the indices.

**Review by akshay-krishnan — CHANGES_REQUESTED** [review]

**Comment by johnwlambert** [comment]
@ayushbaid can you pull master into this branch, so we can check the metrics and see if this is something we would like to do?

**Comment by ayushbaid** [comment]
The main hypothesis behind this PR is: "We have good pose from the averaging module, and hence we can use a prior on them to preserve them and not be corrupted by outlier tracks in bundle adjustment."

Let's compare the angular rotation and translation errors after averaging and after BA in the baseline:

# Door-12 dataset

| Metric | Post-averaging | Post-BA |
| ------- | -------------- | ------ |
| Mean rotation angular error |  0.115 |  0.029 |
| Median rotation angular error  | 0.095 | 0.024 |
| Mean translation angular error |  0.976 |  0.649  |
| Median translation angular error |  0.594 | 0.013 |

# Skydio-32 dataset

| Metric | Post-averaging | Post-BA |
| ------- | -------------- | ------ |
| Mean rotation angular error |   7.95404  |  9.022 |
| Median rotation angular error  | 5.5387 | 8.882 |
| Mean translation angular error |  14.1855 |  8.818  |
| Median translation angular error | 8.085 | 7.1398 |

**Comment by ayushbaid** [comment]
For the new changes, here are the metrics:

# Skydio-32 dataset


| Metric | Post-averaging | Post-BA |
| ------- | -------------- | ------ |
| Mean rotation angular error |   7.95404  |  8.409 |
| Median rotation angular error  | 5.539 | 7.533 |
| Mean translation angular error |  13.9015 |  8.366  |
| Median translation angular error | 7.678| 6.797 |

**Comment by ayushbaid** [comment]
The metrics on the pose errors do not point to any major change. Maybe we need metrics to judge the "correctness" of 3D points too?

**Comment by ayushbaid** [comment]
Not a priority right now.

**Pull request closed without merging**